### PR TITLE
step2 tags

### PR DESCRIPTION
**Pull request by Pun-it**
* Added filters for difficulty type
* Added filtering tags, these are only visible when the corresponding data is available 
image

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### Code Review Summary

#### High-Level Observations:
- **Features Introduced**: Addition of `difficulty` attribute across question models (QuestionPaper) and consistent handling in serialization, API schemas, and UI components.
- **Schema & Data Handling**: Proper injection of difficulty into question objects, mapping during serialization, and validation in tests.
- **UI Enhancements**: Filter menu expanded with difficulty and question type filters; difficulty labels now displayed with color-coded badges.
- **Tests & Validation**: Functional tests confirm `difficulty` defaults and serialization correctness. Frontend tests validate display logic and styling.

#### Issues & Improvements:
1. **Redundant & Missing Types/Defaults**:
   - Default `difficulty="Average"` is set at the model level, but code inconsistently treats difficulty cases (e.g., `'average'`, `'Easy'`). Use a standardized enum (e.g., `DifficultyEnum`) to enforce type safety.
   - Tests check for `"Medium"` as default, but models set `"Average"` — inconsistent. Standardize across all code.

2. **Unnecessary Reassignments & Loops**:
   - In `question_paper_service.py`, post-processing of items sets `item['difficulty']` via a loop, but only if `'difficulty'` key exists. Instead, handle this centrally during serialization or in schema validation.

3. **Schema & Validation Weaknesses**:
   - The modified `schema_dict` injection isn't using enum validation, leading to potential inconsistency. Consider pydantic `constr` or `Enum`.
   - Tests for models assume default `"Medium"`, which conflicts with model default `"Average"`.

4. **Frontend Complexity & Style**:
   - Advertising advanced filter menu UI with multiple nested elements — ensure accessibility, and avoid over-complicated DOM. Simplify toggle logic if needed.
   - Repetitive class bindings could be refactored using helper functions for cleaner code.

5. **Test Coverage & Quality**:
   - Tests are comprehensive for serialization and defaults but could include edge cases (e.g., explicit invalid difficulty).
   - Frontend tests rely on string checks; consider snapshot testing for more robustness.

### Design & Patterns:
- Good use of pydantic for models; move towards `Enum` for `difficulty` instead of arbitrary strings.
- Complex filtering UI is well-structured but could be simplified with reusable components.
- Centralize the `difficulty` normalization in models, serializers, and API schemas to avoid duplication.

### Score: 78/100
- **Strengths**: Correct data integration, clear UI improvements, functional tests.
- **Minor issues**: Slight inconsistency in difficulty defaults, redundant post-processing, opportunity to enforce stricter typing.

**Recommendation**: Approve after implementing:
- `DifficultyEnum` in models and schemas.
- Centralize difficulty normalization at data ingestion and serialization points.
- Clean up filtering logic and test for invalid values.

File: shiksha-api/app-service/app/models/question_paper.py [hunk 1]
```diff
@@ -26,6 +26,7 @@ class QuestionBankMetadata(BaseModel):
 
 class TextQuestion(BaseModel):
     question: str = ""
+    difficulty: str = "Average"
```
Comment (mamuqsit) [inline shiksha-api/app-service/app/models/question_paper.py]: Would be nice to constrain difficulty to expected values.

Per the prompt in question_paper_service.py:
> "`difficulty` should be one of: 'Easy', 'Average', 'Difficult'.\n"

We can define a reusable DifficultyType and use it throughout the 3 models:
```py
from typing import Literal, TypeAlias
from pydantic import BaseModel

DifficultyType: TypeAlias = Literal["Easy", "Average", "Difficult"]

class TextQuestion(BaseModel):
    difficulty: DifficultyType
```

File: shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-generation/question-bank-template/question-bank-template.component.html [hunk 2]
```diff
@@ -14,22 +14,96 @@ <h2 class="text-base md:text-xl font-bold text-content">Step 2/3: Select Questio
...
-            [class.text-white]="filterSource==='Pregenerated Questions'"
-            class="px-2 py-1 rounded border">Pregenerated</button>
+    <div class="p-3 border-b bg-white flex items-center gap-3 relative">
+      <!-- Filter Dropdown Component-like logic -->
+      <div class="relative">
+        <button (click)="toggleFilterMenu()"
+          class="flex items-center gap-2 px-3 py-1.5 rounded-lg border border-gray-200 hover:border-primary active:bg-gray-50 transition-all shadow-sm bg-white">
+          <img src="assets/icons/ri--filter-line.svg" alt="Filter" class="w-4 h-4" />
+          <span class="text-sm font-semibold text-gray-700">Filters</span>
+          <img src="assets/icons/down-arrow.svg" alt="" class="w-2.5 h-2.5 opacity-50 transition-transform"
```
Comment (mamuqsit) [inline shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-generation/question-bank-template/question-bank-template.component.html]: Strange, these icons are "invisible" because they are white in color...

Two approaches, either we:
1. Add `style="filter: invert(1);"` to these `` elements (looks better), or
2. Make button blue - `class="btn-primary"`

**Review by mamuqsit — CHANGES_REQUESTED** [review]

File: shiksha-api/app-service/app/models/question_paper.py (hunk 1 above)
Comment (Pun-it) [reply shiksha-api/app-service/app/models/question_paper.py]: I've updated the models to use a restricted DifficultyType (Literal["Easy", "Average", "Difficult"]) as suggested to ensure that only expected values are allowed.

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary & Recommendations

**Strengths & Improvements:**
- Introduction of `DifficultyType` as a Literal union ("Easy", "Average", "Difficult") significantly enhances type safety.
- Consistent inclusion of `difficulty` across question models and injection into response processing eases data handling.
- Extended serialization tests ensure `difficulty` persists through serialization/deserialization.
- UI/UX adjustments in the filter component improve clarity and usability.

**Issues & Concerns:**
- **Type Safety & Validation**:
  - `difficulty` defaults as `"Average"` but should be validated at model creation for stricter constraints.
  - Some models ('TextQuestion', 'MatchingListQuestion') default `difficulty="Medium"` during instantiation (new code vs. original), risking inconsistency.
- **Redundant & Ill-typed Code**:
  - Post-processing in `_generate_questions_batch()` alters `difficulty` text case, which can be avoided if input is validated upfront.
  - `schema_dict()` now includes `"difficulty"` with default `"Medium"`, diverging from the explicit Literal union, risking inconsistent data.
- **Repeated Logic & Complexity**:
  - `applyFilters()` method in UI is complex; consider using Angular pipes or separate filter services.
  - UI filter toggling code is intricate; could be simplified with reusable components.
- **Testing Scope**:
  - Tests cover serialization and model defaults, but validation concerns (invalid difficulty values) are missing.
  - No coverage for API responses or edge conditions.

### Recommendations:
- Enforce `difficulty` as `Literal["Easy", "Average", "Difficult"]` across all models.
- Remove redundant default assignments like `"Medium"`; rely solely on constraints.
- Validate or coerce `difficulty` values during input or model creation.
- Simplify UI filter logic; consider dedicated filter components or Angular best practices.
- Expand tests to validate invalid `difficulty` inputs and edge cases.

### Score: 75/100
- Satisfactory improvements but needs stricter typing, validation, and code simplification before deployment.

### References:
- [Pydantic Schema Validation](https://pydantic.dev/)
- [Angular Best Practices for Filtering](https://angular.io/guide/pipes)
- [Type Safety with Literal Types](https://mypy.readthedocs.io/en/stable/language_types.html#literal-type)

**Next Steps:**  
- Strengthen validation for `difficulty` across models  
- Simplify complex filter UI logic  
- Extend tests to ensure robustness

File: shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-generation/question-bank-template/question-bank-template.component.html (hunk 2 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-generation/question-bank-template/question-bank-template.component.html]: Went with your first approach and added style="filter: invert(1);" to the  elements, which makes them cleanly visible without altering the button's overall color scheme

**Review by mamuqsit — APPROVED** [review]
